grails-app/services/org/pih/warehouse/invoice/PrepaymentInvoiceService.groovy

```diff
@@ -23,47 +57,150 @@ class PrepaymentInvoiceService {
+                invoiceItemType: InvoiceItemType.REGULAR
+        )
+        invoiceItem.addToOrderItems(orderItem)
+        return invoiceItem
+    }
 
-        if (order?.orderItems?.empty && order?.orderAdjustments?.empty) {
-            throw new Exception("No order items or order adjustments found for given order")
-        }
+    InvoiceItem createFromShipmentItem(ShipmentItem shipmentItem) {
+        OrderItem orderItem = shipmentItem.orderItems?.find { it }
+        Integer quantity = shipmentItem.quantityToInvoiceInStandardUom ? (shipmentItem.quantityToInvoiceInStandardUom / orderItem?.quantityPerUom) : 0
+        InvoiceItem invoiceItem = new InvoiceItem(
+                // InvoiceItem quantity is in UoM not in standard UoM
+                quantity: quantity,
+                product: shipmentItem.product,
+                glAccount: shipmentItem.product.glAccount,
+                budgetCode: orderItem?.budgetCode,
+                quantityUom: orderItem?.quantityUom,
+                quantityPerUom: orderItem?.quantityPerUom ?: 1,
+                unitPrice: orderItem?.unitPrice,
+                // For shipment items invoice item amount is equal to (shipped item quantity * unit price) in UoM quantity
+                amount: quantity * orderItem?.unitPrice,
+                invoiceItemType: InvoiceItemType.REGULAR
+        )
+        invoiceItem.addToShipmentItems(shipmentItem)
+        return invoiceItem
+    }
 
-        order.activeOrderItems.each { OrderItem orderItem ->
-            InvoiceItem invoiceItem = invoiceService.createFromOrderItem(orderItem)
-            invoice.addToInvoiceItems(invoiceItem)
-        }
+    InvoiceItem createFromOrderAdjustment(OrderAdjustment orderAdjustment) {
+        InvoiceItem invoiceItem = new InvoiceItem(
+                budgetCode: orderAdjustment.budgetCode,
+                product: orderAdjustment.orderItem?.product,
+                glAccount: orderAdjustment.glAccount ?: orderAdjustment.orderItem?.glAccount ?: orderAdjustment.orderAdjustmentType?.glAccount,
+                quantity: orderAdjustment?.canceled ? 0 : 1,
+                quantityUom: orderAdjustment.orderItem?.quantityUom,
+                quantityPerUom: orderAdjustment.orderItem?.quantityPerUom ?: 1,
+                unitPrice: orderAdjustment.totalAdjustments,
+                // For non-canceled order adjustments amount is equal to the total adjustments
+                amount: orderAdjustment?.canceled ? 0 : orderAdjustment.totalAdjustments,
+                invoiceItemType: InvoiceItemType.REGULAR
+        )
+        invoiceItem.addToOrderAdjustments(orderAdjustment)
+        return invoiceItem
+    }
 
-        order.activeOrderAdjustments.each { OrderAdjustment orderAdjustment ->
-            InvoiceItem invoiceItem = invoiceService.createFromOrderAdjustment(orderAdjustment)
-            invoice.addToInvoiceItems(invoiceItem)
+    InvoiceItem createInverseItemForOrderItem(OrderItem orderItem) {
```

Comment:
if this is only for items with no shipments (ie cancelled ones), it's probably best to be clear in the method name. `createInverseItemForCancelledOrderItem()` or something